### PR TITLE
Report infinite tracing (and other) configuration in connect/agent_settings payload

### LINKED ISSUE #1165: Report infinite tracing configs on Connect
The .NET Agent current does not report any infinite tracing related configs in the Connect payload. We should consider having the agent to report these for supportability purposes.

### PR DESCRIPTION
**Pull request by JcolemanNR**
## Description
Closes #1165.

This PR introduces a new json serializable configuration reporting class that implements `IConfiguration`. This class is then included in the `connect` and `agent_settings` payload sent to New Relic. The goals are to improve supportability by GTS, and make this data queryable in BMDS.

After getting approval from GTS, some existing json names were changed.

This PR needs:
- [x] More refinement of json names/grouping
- [x] Some config elements can be ignored as they are redundant
- [x] At least one unit test that completely mocks IConfiguration for safety

# Author Checklist
- [x] Unit tests, Integration tests, and Unbounded tests completed
- [x] Performance testing completed with satisfactory results (if required)
- [x] [Agent Changelog](/src/Agent/CHANGELOG.md) or [Lambda Agent changelog](/src/AwsLambda/CHANGELOG.md) updated 

# Reviewer Checklist
- [ ] Perform code review
- [ ] Pull request was adequately tested (new/existing tests, performance tests)
- [ ] Review Changelog

image

### CODE REVIEW

File: src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs [hunk 1]
```diff
@@ -1,114 +1,607 @@
...
     {
-        [JsonProperty("agent")]
+        private readonly IConfiguration _configuration;
+        public ReportedConfiguration(IConfiguration configuration)
+        {
+            _configuration = configuration;
+        }
+
+        //[JsonProperty("agent.name")]
+        [JsonProperty("name")]
```
Comment (nr-ahemsath) [inline src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs:26]: So, this value (".NET Agent") was being reported under the key name "agent" and will now be reported under key name "name".  Is there any concern that this change might break anything?

I don't see anything in BMDS that looks like a problem.

File: src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs [hunk 2]
```diff
@@ -1,114 +1,607 @@
...
+        public object AgentRunId => _configuration.AgentRunId?.ToString();
 
-        [JsonProperty("encoding_key")]
-        public string EncodingKey { get; set; }
+        [JsonProperty("agent.enabled")]
+        public bool AgentEnabled => _configuration.AgentEnabled;
 
-        [JsonProperty("trusted_account_ids")]
-        public IEnumerable<long> TrustedAccountIds { get; set; }
+        //[JsonProperty("agent.license_key")]
```
Comment (nr-ahemsath) [inline src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs:37]:
```suggestion
```

File: src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs [hunk 3]
```diff
@@ -1,114 +1,607 @@
...
+        [JsonProperty("error_collector.ignore_messages")]
+        public IDictionary<string, IEnumerable<string>> IgnoreErrorMessagesForAgentSettings => _configuration.IgnoreErrorMessagesForAgentSettings;
+
+        [JsonProperty("agent.request_headers_map")]
+        public Dictionary<string, string> RequestHeadersMap => _configuration.RequestHeadersMap;
+
+        // Use old name
+        //[JsonProperty("agent.encoding_key")]
+        [JsonProperty("encoding_key")]
+        public string EncodingKey => _configuration.EncodingKey;
```
Comment (nr-ahemsath) [inline src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs:309]: What is this encoding key, and is it something we're concerned about leaking?

File: src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs [hunk 4]
```diff
@@ -1,114 +1,607 @@
...
+        [JsonProperty("browser_monitoring.auto_instrument")]
+        public bool BrowserMonitoringAutoInstrument => _configuration.BrowserMonitoringAutoInstrument;
+
+        [JsonProperty("browser_monitoring.beacon_address")]
+        public string BrowserMonitoringBeaconAddress => _configuration.BrowserMonitoringBeaconAddress;
+
+        [JsonProperty("browser_monitoring.error_beacon_address")]
+        public string BrowserMonitoringErrorBeaconAddress => _configuration.BrowserMonitoringErrorBeaconAddress;
+
+        //[JsonProperty("browser_monitoring.javascript_agent")]
```
[the remark this answers is not in the document]
Comment (JcolemanNR) [reply src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs:62]: changed in my working branch

File: src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs (hunk 3 above)
Comment (JcolemanNR) [reply src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs:309]: It looks like this has to do with CAT functionality... and it was sent up by the previous 'ReportedConfiguration', so apparently it's not a security concern :)

File: src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs [hunk 5]
```diff
@@ -1,114 +1,607 @@
...
+        [JsonProperty("proxy.host")]
+        public string ProxyHost => _configuration.ProxyHost;
+
+        [JsonProperty("proxy.uri_path")]
+        public string ProxyUriPath => _configuration.ProxyUriPath;
+
+        [JsonProperty("proxy.port")]
+        public int ProxyPort => _configuration.ProxyPort;
+
+        //[JsonProperty("proxy.username")]
```
[the remark this answers is not in the document]
Comment (JcolemanNR) [reply src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs:356]: In general, I kept them as places that the existing names conflicted with my naming scheme, or for values I wasn't sure needed to be ignored. I'll clean them all up and resolve things.

File: src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs [hunk 6]
```diff
@@ -1,114 +1,607 @@
...
+        public IEnumerable<string> ExpectedErrorClassesForAgentSettings => _configuration.ExpectedErrorClassesForAgentSettings;
 
         [JsonProperty("error_collector.expected_messages")]
-        public IDictionary<string, IEnumerable<string>> ErrorCollectorExpectedMessages { get; set; }
+        public IDictionary<string, IEnumerable<string>> ExpectedErrorMessagesForAgentSettings => _configuration.ExpectedErrorMessagesForAgentSettings;
 
-        [JsonProperty("error_collector.expected_status_codes")]
-        public IEnumerable<string> ErrorCollectorExpectedStatusCodes { get; set; }
+        // TODO: Less snark
+        [JsonProperty("error_collector.expected_status_codes_for_agent_settings_trolololol")]
```
[the remark this answers is not in the document]
Comment (JcolemanNR) [reply src/Agent/NewRelic/Agent/Core/Configuration/ReportedConfiguration.cs:288]: In this case, there were two properties that roughly presented the same information, so there was a naming collision. I have selected this one as the winner as it presents the data specifically (and only) 'ForAgentSettings' :)

**Pull request merged**